### PR TITLE
RP2 - PIO module

### PR DESCRIPTION
**Pull request by NitiKaur**
1st draft for PIO module- brief overview with explanation of each command

### CODE REVIEW

**Comment by dpgeorge** [comment]
This is a good start, and the general overview here is well suited to a tutorial or general introduction section.

The specific commands and the details of their arguments are more of a reference and I think would be better placed in a new section at the end of `docs/library/rp2.rst`.

File: docs/rp2/pio.rst [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    see sec 3.4.7 of RPi docs for details
+
+mov(dist, src)
+	dist : pin, X, Y, ISR, OSR, PC (all 0-31)
+	src : pin, X, Y, ISR, OSR (all 0-31)
+    see sec 3.4.8 of RPi docs for details
+
+irq(mod, index = None)
+	index : IRQ (0-7)
+    see sec 3.4.9 of RPi docs for details
```
Comment (dpgeorge) [inline docs/rp2/pio.rst:71]: These are the following forms of this instruction (`index` is an integer value):
- `irq(index)`
- `irq(rel(index))`
- `irq(block, index)`
- `irq(block, rel(index))`
- `irq(clear, index)`
- `irq(clear, rel(index))`
- `irq(block | clear, index)`
- `irq(block | clear, rel(index))`

That can be summarised as 2 forms:
- form 1: `irq(index)`
- form 2: `irq(mode, index)`
where:
- `index` can be an integer or `rel(integer)`
- `mode` can be: `block` or `clear` or `block | clear`.

Allowed IRQ numbers are 0-7 (0-3 are visible from to the processor, 4-7 are internal to the state machines).

File: docs/rp2/pio.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+self(dest, data)
```
Comment (dpgeorge) [inline docs/rp2/pio.rst:73]: should be "set"

**Comment by mattytrentini** [comment]
This is already of use and close to what we asked of you @NitiKaur. But the more I dig into the PIO programming (I'm quite new to it), the more I think we need more, of which, this reference will form a component.

In particular:

- A description of `rp2.StateMachine`
  - How many, how to start, what methods etc
- How `nop()` works
- How `rel()` affects indices
- An explanation for `wrap_target()`/`wrap()`
- How labels are implemented
- Methods on some of the instructions (`.side()` for example) 
- How IRQ can map to interrupts

In short, enough documentation so that _all_ of the techniques used in the PIO [examples](https://github.com/micropython/micropython/tree/master/examples/rp2) are covered.

We'll discuss this further at the next meeting.

File: docs/library/rp2.rst [hunk 3]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
     This exception is raised from `asm_pio()` or `asm_pio_encode()` if there is
     an error assembling a PIO program.
 
+Each state machine has a set of instructions- which allow for a wide range of
```
Comment (dpgeorge) [inline docs/library/rp2.rst:74]: This section could do with a heading, for example "PIO assembly language"

File: docs/library/rp2.rst [hunk 4]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+
+wait(polarity, src, index)
+     polarity: 0-1
+     src: gpio, pin, irq
+     index: 0-31
+     wait for high/low on pin 0-31 or irq 0-31
+     see sec 3.4.3 of RPi docs for details
+
+in(src, data)
+	src : pin, X, Y, ISR, OSR(all 0-31)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:93]: these all need to be lower case: x, y, isr, osr.  also it's `in_` to not clash with the `in` keyword

File: docs/library/rp2.rst [hunk 5]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+	data : bit count(0-31)
+    see sec 3.4.4 of RPi docs for details
+
+out(dist,data)
+	dist : pin, X, Y, PINDIRS, AC, ISR, OSR(all 0-31)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:98]: dist should be dest, and lowercase for x, y, pindirs, isr, osr (AC is not supported I don't think)

File: docs/library/rp2.rst [hunk 6]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+    see sec 3.4.6 of RPi docs for details
+
+pull(value = 0, value = 2)
+	pull(if empty) : 1 (default = 0)
+	block : 1 (default = 1)  
+	no-block = 0
+    see sec 3.4.7 of RPi docs for details
+
+mov(dist, src)
+	dist : pin, X, Y, ISR, OSR, PC (all 0-31)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:114]: dist should be dest (for destination), and lowercase for the options

File: docs/library/rp2.rst [hunk 7]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+	src : pin, X, Y, ISR, OSR (all 0-31)
+    see sec 3.4.8 of RPi docs for details
+
+set(dest, data)
+	dest : X, Y, PINS, PINDIRS
```
Comment (dpgeorge) [inline docs/library/rp2.rst:119]: lowercase

File: docs/library/rp2.rst [hunk 8]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+irq(mod, index = None)
+	These are the following forms of this instruction(index is an integer value 
+	from 0-7)
+        * irq(index)
+	* irq(rel(index))
+	* irq(block, index)
+	* irq(block, rel(index))
+	* irq(clear, rel(index))
+	* irq(block | clear, index)
+	* irq(block | clear, rel(index))
```
Comment (dpgeorge) [inline docs/library/rp2.rst:132]: either keep this set of possible ways of writing it, or the below set of 2 forms, but not both

File: docs/rp2/pio.rst [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+We can take the pio_1hz example for a sinple understanding. We are creating an
+object of class StateMachine which will display the output on pin 25. The
+'blink_1hz' program uses the PIO to blink an LED and raise the IRQ at 1 HZ.
+Turn the LED on with the help of the set instruction. Set value 31 on register
+X. Then wait for 30 cycles with the help of nop() instruction. Then with the
+help of jmp, go to the code part which has label as 'delay high' for the
+instruction offset. The next part of the code teaches us to turn the LED off
+by turning the LED off with the help of set instruction. Here the label is
+'delay_low' for the instruction offset part of the jmp instruction.
```
Comment (dpgeorge) [inline docs/rp2/pio.rst:55]: It would be good to put the code for blink_1hz here in the docs so readers can easily see what is going on

File: docs/rp2/pio.rst [hunk 10]
```diff
@@ -0,0 +1,59 @@
+.. _rp2_pio:
```
Comment (dpgeorge) [inline docs/rp2/pio.rst:1]: rename this file to `docs/rp2/tutorial/pio.rst`

File: docs/library/rp2.rst [hunk 11]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+	dist : pin, X, Y, ISR, OSR, PC (all 0-31)
+	src : pin, X, Y, ISR, OSR (all 0-31)
+    see sec 3.4.8 of RPi docs for details
+
+set(dest, data)
+	dest : X, Y, PINS, PINDIRS
+	data : value (0-31)
+    see sec 3.4.10 of RPi docs for details
+
+irq(mod, index = None)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:123]: instead of this line, I'd suggest breaking this instruction into 2 parts:
```
irq(index)
     index: <integer> or rel(<integer>)

irq(mode, index)
    mode: block, clear, block|clear
    index: <integer> or rel(<integer>)
```

File: docs/library/rp2.rst [hunk 12]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
     This exception is raised from `asm_pio()` or `asm_pio_encode()` if there is
     an error assembling a PIO program.
 
+Each state machine has a set of instructions- which allow for a wide range of
+behaviourial or location changes to the data. They are written in assembly
+language which is later used in a MicroPython program to perform specific
+tasks. These instructions are -
+
+jmp(cond, lebel=None)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:79]:
```
jmp(label)
    label: label to jump to unconditionally

jmp(cond, label)
    cond: not_x, x_dec, not_y, y_dec, x_not_y, pin, not_osre
    label: label to jump to if condition is true
```

File: docs/library/rp2.rst [hunk 13]
```diff
@@ -71,6 +71,79 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+	src : pin, X, Y, ISR, OSR(all 0-31)
+	data : bit count(0-31)
+    see sec 3.4.4 of RPi docs for details
+
+out(dist,data)
+	dist : pin, X, Y, PINDIRS, AC, ISR, OSR(all 0-31)
+	data : bit count(0-31)
+    see sec 3.4.5 of RPi docs for details
+
+push(value = 0, value = 2)
```
Comment (dpgeorge) [inline docs/library/rp2.rst:102]:
```
push()
push(block)
push(noblock)
push(iffull)
push(iffull, block)
push(iffull, noblock)
```

File: docs/library/rp2.rst (hunk 11 above)
Comment (dpgeorge) [reply docs/library/rp2.rst:123]: or it might be clearer to just write out all 7 possible options explicitly, as done below

**Comment by dpgeorge** [comment]
There are also "meta instructions":
- wrap_target
- wrap
- label
- word
- nop

Then also `side()` and `[x]` (delay).  They can be applied to all (?) instructions by writing them after the instruction, eg:
```
<instruction>  .side(<integer>)  [<integer_delay>]
```
You can apply side or delay or both or neither.

File: docs/rp2/tutorial/pio.rst [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+protocols with standard devices. For devices where we do not have hardware
+support for such protocols or where there is a requirement of more I/O
+channels, PIO(Programmable Input Ouput) comes intp play. Also, typical
+micropython devices make use of a technique called bit banging in which pins
+are rapidly turned on and off to transmit data. This makes the entire process
+very slow as the processor concentrates on bit banging rather than exceuting
+any other process. However, PIO allows it to carry on with the bit banging
+process while the processor resources are used for executing some other work. 
+
+Along with the two main cotex - MO+ processing cores, RP2040 has two PIO
```
Comment (peterhinch) [inline docs/rp2/tutorial/pio.rst:13]: sp: cortex

File: docs/rp2/tutorial/pio.rst [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+blocks each of which has four state machines. These state machines take
+process data in FIFO (First-In-First-Out) format. When these state machines
+are free, they will put in data from the queue and execute the instruction.
+This way, the PIO state machines and the main process works simultaneously
+without being much dependant on the other.
```
Comment (peterhinch) [inline docs/rp2/tutorial/pio.rst:18]: sp: dependent

**Comment by dpgeorge** [comment]
The 4 directives that MicroPython supports are: `wrap`, `wrap_target`, `word` and `label`

File: docs/rp2/tutorial/pio.rst [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+* PULL : recieves data from TX FIFO
+
+* MOV : moves data from a souce to a destination
+
+* IRQ : sets or clears the input flag
+
+* WAIT : pauses until a particular action happens
+
+* JMP : moves to some target locations in differnt parts of code.
```
Comment (dpgeorge) [inline docs/rp2/tutorial/pio.rst:42]: missing SET

File: docs/library/rp2.rst [hunk 17]
```diff
@@ -71,6 +71,90 @@ For running PIO programs, see :class:`rp2.StateMachine`.
...
+      * index can be an integer or rel(integer)
+      * mode can be: block or clear or block | clear
+      
+    Allowed IRQ numbers are 0-7 (0-3 are visible from to the processor, 4-7 are 
+    internal to the state machines).          
+   see sec 3.4.9 of RPi docs for details
+
+nop(self)
+    see sec 3.3.7 of RPi docs for details
+
```
Comment (dpgeorge) [inline docs/library/rp2.rst:157]: Also need to add here the delay and side-set modifiers.

Comment (dpgeorge) [reply docs/library/rp2.rst:157]: This still needs to be done (describing delay and side-set).  Also list the 4 directives in this reference.

**Comment by kevindawson** [comment]
@NitiKaur 

Our old friend:

- Subject line should match '^[^!]+: [A-Z]+.+ .+\\.$':

rears it's head again

[see commit-conventions](https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md#git-commit-conventions)

**Review by alphaFred — CHANGES_REQUESTED** [review]
Found some typos

**Review by alphaFred — APPROVED** [review]

**Review by alphaFred — APPROVED** [review]

**Review by alphaFred — APPROVED** [review]

**Comment by NitiKaur** [comment]
@dpgeorge  
I am getting the following error while building docs. 
/home/runner/work/micropython/micropython/docs/rp2/tutorial/pio.rst: WARNING: document isn't included in any toctree
I tried copying the ESP8266 file structure as suggested for the ESP32 but it didn't work out as desired. It will be great if you can tell me how I should include this index file

**Comment by alphaFred** [comment]
For now you have `tutorial/intro.rst` which is referenced in `quickref.rst` but you have not referenced your `pio.rst`anywhere as far as I could see.
When you check out the `pyboard` documentation there is a separated document called ` micropython/docs/pyboard/tutorial/index.rst` referencing all tutorial documents. This would be a good option, especially when adding more tutorials in the future I guess.
